### PR TITLE
[FEATURE] IntelliJ Syntax Highlighting Plugin

### PR DESCRIPTION
**Pull request by Tigermouthbear**
This adds a IntelliJ syntax highlighting plugin to the extensions folder (`extensions/intellij`). I have requested the plugin to be published through the JetBrains marketplace and it should be approved in the next couple of days (https://plugins.jetbrains.com/plugin/18693-melody).

I plan on adding code completion to the plugin in the near future too.

### CODE REVIEW

**Comment by yoav-lavi** [comment]
This is awesome! Will go over the PR soon.

**Comment by yoav-lavi** [comment]
@Tigermouthbear the licenses here seem to be under your name, I'm not completely sure how this is normally handled when a repo has a contribution with its own license. Perhaps we could use Melody as the name?

Will respond to the rest soon, thank you!

**Comment by Tigermouthbear** [comment]
I can change the license to match Melodys if you'd prefer not to have other licensed code in the repo.

File: extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    space="<space>"
+    newline="<newline>"
+    tab="<tab>"
+    return="<return>"
+    feed="<feed>"
+    null="<null>"
+    digit="<digit>"
+    vertical="<vertical>"
+    word="<word>"
+    alphabet="<alphabet>"
```
Comment (yoav-lavi) [inline extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf:44]: @Tigermouthbear This changed to `alphabetic` recently and a few new symbols were added

File: extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    capture="capture"
+    some="some"
+    match="match"
+    over="over"
+    option="option"
+    not="not"
+    either="either"
+    any="any"
+    ahead="ahead"
+    behind="behind"
```
Comment (yoav-lavi) [inline extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf:28]: `let` and `lazy` are also keywords

**Review by yoav-lavi — CHANGES_REQUESTED** [review]
Looks good! left a few comments regarding the syntax. Would it be possible to add this to the language ignores in the .gitattributes? Thanks!

File: extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf (hunk 1 above)
Comment (yoav-lavi) [reply extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf:44]: There are also a few more symbols added since these

**Comment by yoav-lavi** [comment]
Regarding the icon - not yet, I will later on

**Comment by yoav-lavi** [comment]
@Tigermouthbear attaching the SVGs here
melody
melody-dark

**Comment by yoav-lavi** [comment]
@Tigermouthbear let me know if you need a version without the text for file icons

**Comment by Tigermouthbear** [comment]
Thanks! I was just using the regex file icon because it fits into the ide theme nice.

**Comment by Tigermouthbear** [comment]
I made the changes you suggested and updated the plugin logo! I released v0.2 to the jetbrains marketplace and it should be available in the next couple of days.

Heres the full changelog: 
            <ul>
                <li>Added let and lazy keyword support</li>
                <li>Added &lt;backspace&gt;, &lt;boundary&gt;, and &lt;alphanumeric&gt; support</li>
                <li>Renamed &lt;alphabet&gt; to &lt;alphabetic&gt;</li>
                <li>Added multi-line comment support</li>
                <li>Added character range support</li>
                <li>Added .melody file extension support</li>
                <li>Switch to new Melody logo!</li>
            </ul>

**Comment by Tigermouthbear** [comment]
also, it looks like the extensions folder is already added to the langauge ignores in .gitattributes? I've never used language ignores but it looks like its already ignored

File: extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    number="regexp:[0-9]+"
+
+    // comments
+    comment="regexp://.*|/\*!(\*/)*\*/|/\*\*/"
+
+    // strings
+    string = "regexp:\"[^\"]*\"|'[^']*'|`[^`]*`"
+
+    // other patterns
+    character = "regexp:\S"
```
Comment (yoav-lavi) [inline extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf:61]: character literals (for negative classes) only support these at the moment: `('a'..'z'|'A'..'Z'| '0'..'9' | "_" | "\\" | "@" | "*" | "$" | "#" | "&" | "^" | "!" | "%")`

**Review by yoav-lavi — CHANGES_REQUESTED** [review]
Looks good except the one comment, thank you!

File: extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf (hunk 3 above)
Comment (Tigermouthbear) [reply extensions/intellij/src/main/kotlin/dev/tigr/melody/plugin/Melody.bnf]: Ok, I just pushed those changes!

**Comment by yoav-lavi** [comment]
Looks great, thank you for your contribution!

**Comment by yoav-lavi** [comment]
@Tigermouthbear it seems like the extension icon is now white on white: 
Screen Shot 2022-03-20 at 16 32 48

**Comment by yoav-lavi** [comment]
Also, would it be possible to add permissions for myself to edit / upload the extension? Thanks!

**Comment by Tigermouthbear** [comment]
oh, im an idiot. I accidentally put the white logo as the light theme icon and the black logo as the dark theme icon. I'll open a new PR and update the plugin. As for allowing you to edit, Id have to create an organization which would then allow you to edit. For this I would need the email and website you want to use for it.

**Pull request merged**